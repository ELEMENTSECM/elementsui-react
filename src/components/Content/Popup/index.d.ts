import * as React from "react";

export interface PopupTargetNode {
	clientHeight?: number;
	clientWidth?: number;
	getBoundingClientRect?: (...args: any[]) => any;
}

export type PopupPlacement = string | boolean;

export interface PopupProps {
	/**
	 * Whether popup can be dragged on the screen. Default: true
	 */
	isDraggable?: boolean;
	/**
	 * CSS class for the <div> that hosts popup and backdrop(overlay). Default: "popup-dialog"
	 */
	hostClassName?: string;
	/**
	 * CSS class for the top <div> of the popup. Default: "popup-container"
	 */
	containerClassName?: string;
	/**
	 * CSS class for the backdrop (overlay) <div>. Default: "popup-overlay"
	 */
	backdropClassName?: string;
	/**
	 * CSS selector for the HTML element serving as a handle of draggable popup. Default: ".popup-container"
	 */
	handle?: string;
	/**
	 * Whether popup should be made visible. Default: true
	 */
	show?: boolean;
	/**
	 * Target node used for positioning of popup relative to it. Popper.js lib used behind the scene.
	 */
	targetNode?: PopupTargetNode;
	/**
	 * One of the placements supported by Popper.js relative to target node. If boolean 'false' value is passed, Popper positioning will not be applied.
	 */
	placement?: PopupPlacement;
	/**
	 * Handler that is called when backdrop is clicked or ESC key pressed. Typically in this handler you will update state so that popup is removed.
	 */
	onHide?: (...args: any[]) => any;
	/**
	 * Standard ARIA attribute that will be set on popup root element
	 */
	"aria-labelledby"?: string;
	/**
	 * Should modal become focused on render
	 */
	autoFocus?: boolean;
}

export default class Popup extends React.Component<PopupProps, any> {
	render(): JSX.Element;
}
