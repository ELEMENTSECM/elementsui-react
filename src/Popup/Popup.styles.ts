export const styles = {
	backdrop: { zIndex: 2000 },
	dialog: {
		zIndex: 2000,
		opacity: 1
	},
	centerToScreen: {
		top: "50%",
		left: "50%",
		transform: "translate(-50%, -50%)"
	}
};
