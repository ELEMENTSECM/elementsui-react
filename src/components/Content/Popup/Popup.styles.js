export const styles = {
	backdrop: { zIndex: 2000 },
	container: {
		zIndex: 2001,
		opacity: 1
	},
	centerToScreen: {
		top: "50%",
		left: "50%",
		transform: "translate(-50%, -50%)"
	}
};
