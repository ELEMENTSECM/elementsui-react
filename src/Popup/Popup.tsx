import * as React from "react";
import { PositionedDraggable } from "react-draggable-elements";
import { Modal } from "react-overlays";
import { Popper } from "react-popper";
import { ReferenceObject } from "popper.js";
import styled from "styled-components";
import { isNil } from "lodash";
import { styles } from "./Popup.styles";

type PopupTargetNode = {
	clientHeight: number;
	clientWidth: number;
	getBoundingClientRect: () => ClientRect | DOMRect;
};

export interface PopupProps {
	/** Whether popup can be dragged on the screen. Default: true */
	isDraggable?: boolean;
	/** CSS class for the <div> that hosts popup and backdrop(overlay). Default: "popup-dialog" */
	hostClassName?: string;
	/** CSS class for the top <div> of the popup. Default: "popup-container" */
	containerClassName?: string;
	/** CSS class for the backdrop (overlay) <div>. Default: "popup-overlay" */
	backdropClassName?: string;
	/** CSS selector for the HTML element serving as a handle of draggable popup. Default: ".popup-container" */
	handle?: string;
	/** Whether popup should be made visible. Default: true */
	show?: boolean;
	/** Target node used for positioning of popup relative to it. Popper.js lib used behind the scene. */
	targetNode?: PopupTargetNode | HTMLElement | null;
	/** One of the placements supported by Popper.js relative to target node. If boolean 'false' value is passed, Popper positioning will not be applied. */
	placement?:
		| "auto-start"
		| "auto"
		| "auto-end"
		| "top-start"
		| "top"
		| "top-end"
		| "right-start"
		| "right"
		| "right-end"
		| "bottom-end"
		| "bottom"
		| "bottom-start"
		| "left-end"
		| "left"
		| "left-start"
		| boolean;
	/** Handler that is called when backdrop is clicked or ESC key pressed. Typically in this handler you will update state so that popup is removed. */
	onHide: (event?: any) => any;
	/** Standard ARIA attribute that will be set on popup root element */
	"aria-labelledby": string;
	/** Should modal become focused on render */
	autoFocus?: boolean;
}

export default class Popup extends React.PureComponent<PopupProps> {
	static defaultProps = {
		isDraggable: true,
		hostClassName: "popup-dialog",
		containerClassName: "popup-container",
		backdropClassName: "popup-overlay",
		handle: ".popup-container",
		show: true,
		targetNode: undefined,
		placement: undefined,
		onHide: undefined,
		"aria-labelledby": undefined,
		autoFocus: true
	};

	onHide = () => {
		const { onHide } = this.props;
		onHide && onHide();
	};

	render() {
		const {
			handle,
			children,
			"aria-labelledby": ariaLabelledBy,
			isDraggable,
			show,
			hostClassName,
			containerClassName,
			backdropClassName,
			targetNode,
			placement,
			autoFocus
		} = this.props;

		return (
			<Modal
				aria-labelledby={ariaLabelledBy}
				backdropClassName={backdropClassName}
				backdropStyle={{ zIndex: 2000 }}
				onHide={this.onHide}
				show={show}
				autoFocus={autoFocus}
				className={hostClassName}
			>
				<Popper referenceElement={targetNode as ReferenceObject} placement={placement as any}>
					{({ ref, style }) => (
						<PositionedDraggable handle={handle} disabled={!isDraggable} bounds="html">
							<div
								ref={ref}
								className={containerClassName}
								style={{
									...styles.container,
									...placement === false ? {} : targetNode ? style : styles.centerToScreen
								}}
							>
								{children}
							</div>
						</PositionedDraggable>
					)}
				</Popper>
			</Modal>
		);
	}
}
