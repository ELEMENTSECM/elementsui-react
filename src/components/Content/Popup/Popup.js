import * as React from "react";
import { PositionedDraggable } from "react-draggable-elements";
import { Modal } from "react-overlays";
import { Popper } from "react-popper";
import { styles } from "./Popup.styles";
import PropTypes from "prop-types";

export default class Popup extends React.PureComponent {
	constructor(props) {
		super(props);
	}

	onHide = () => {
		const { onHide } = this.props;
		onHide && onHide();
	};

	render() {
		const {
			handle,
			children,
			"aria-labelledby": ariaLabelledBy,
			isDraggable,
			show,
			hostClassName,
			containerClassName,
			backdropClassName,
			targetNode,
			placement,
			autoFocus,
			manualPlacement
		} = this.props;

		return (
			<Modal
				aria-labelledby={ariaLabelledBy}
				backdropClassName={backdropClassName}
				backdropStyle={styles.backdrop}
				onHide={this.onHide}
				show={show}
				autoFocus={autoFocus}
				className={hostClassName}
			>
				<Popper referenceElement={targetNode} placement={placement}>
					{({ ref, style }) => (
						<PositionedDraggable handle={handle} disabled={!isDraggable} bounds="html">
							<div
								ref={ref}
								className={containerClassName}
								style={{
									...styles.container,
									...(manualPlacement ? {} : targetNode ? style : styles.centerToScreen)
								}}
							>
								{children}
							</div>
						</PositionedDraggable>
					)}
				</Popper>
			</Modal>
		);
	}
}

Popup.defaultProps = {
	isDraggable: true,
	hostClassName: "popup-dialog",
	containerClassName: "popup-container",
	backdropClassName: "popup-overlay",
	handle: ".popup-container",
	show: true,
	targetNode: undefined,
	placement: undefined,
	onHide: undefined,
	"aria-labelledby": undefined,
	autoFocus: true
};

Popup.propTypes = {
	/** Whether popup can be dragged on the screen. Default: true */
	isDraggable: PropTypes.bool,
	/** CSS class for the <div> that hosts popup and backdrop(overlay). Default: "popup-dialog" */
	hostClassName: PropTypes.string,
	/** CSS class for the top <div> of the popup. Default: "popup-container" */
	containerClassName: PropTypes.string,
	/** CSS class for the backdrop (overlay) <div>. Default: "popup-overlay" */
	backdropClassName: PropTypes.string,
	/** CSS selector for the HTML element serving as a handle of draggable popup. Default: ".popup-container" */
	handle: PropTypes.string,
	/** Whether popup should be made visible. Default: true */
	show: PropTypes.bool,
	/** Target node used for positioning of popup relative to it. Popper.js lib used behind the scene. */
	targetNode: PropTypes.shape({
		clientHeight: PropTypes.number,
		clientWidth: PropTypes.number,
		getBoundingClientRect: PropTypes.func
	}),
	/** One of the placements supported by Popper.js relative to target node. */
	placement: PropTypes.string,
	/** Handler that is called when backdrop is clicked or ESC key pressed. Typically in this handler you will update state so that popup is removed. */
	onHide: PropTypes.func,
	/** Standard ARIA attribute that will be set on popup root element */
	"aria-labelledby": PropTypes.string,
	/** Should modal become focused on render */
	autoFocus: PropTypes.bool,
	/** Disable automatic positioning when popup is opened */
	manualPlacement: PropTypes.bool
};
