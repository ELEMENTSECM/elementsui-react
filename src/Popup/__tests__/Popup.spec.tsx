import { shallow } from "enzyme";
import Popup from "../index";

describe("Popup", () => {
	test("displays correctly", () => {
		const wrapper = shallow(
			<Popup>
				<span>popup content</span>
			</Popup>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
