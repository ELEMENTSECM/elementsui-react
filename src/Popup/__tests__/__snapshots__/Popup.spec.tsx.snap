// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup displays correctly 1`] = `
<Modal
  autoFocus={true}
  className="popup__dialog"
  renderBackdrop={[Function]}
  renderDialog={[Function]}
  show={true}
  style={
    Object {
      "left": "50%",
      "opacity": 1,
      "top": "50%",
      "transform": "translate(-50%, -50%)",
      "zIndex": 2000,
    }
  }
/>
`;
